### PR TITLE
Updated Raster classes to throw an exception when 'StbImage load' fai…

### LINKED ISSUE #1468: When 'StbImage load' fails to load an image it returns Null which is left unhandled
The suggestion is to raise an exception if the Image failed to load.

### PR DESCRIPTION
**Pull request by erikhagglund**
…ls to open the image

Fixes #1468
@emilwestergren

### CODE REVIEW

File: source/draw/RasterImage.ooc [hunk 1]
```diff
@@ -59,6 +59,9 @@ RasterImage: abstract class extends Image {
 	open: static func ~unknownType (filename: String) -> This {
 		x, y, imageComponents: Int
 		data := StbImage load(filename, x&, y&, imageComponents&, 0)
+		if (data == null) {
```
Comment (emilwestergren) [inline source/draw/RasterImage.ooc]: You don't need brackets here.

File: source/draw/RasterMonochrome.ooc [hunk 2]
```diff
@@ -198,6 +198,9 @@ RasterMonochrome: class extends RasterPacked {
 		x, y, imageComponents: Int
 		requiredComponents := 1
 		data := StbImage load(filename, x&, y&, imageComponents&, requiredComponents)
+		if (data == null) {
```
Comment (emilwestergren) [inline source/draw/RasterMonochrome.ooc]: Same here

File: source/draw/RasterRgba.ooc [hunk 3]
```diff
@@ -153,6 +153,9 @@ RasterRgba: class extends RasterPacked {
 		x, y, imageComponents: Int
 		requiredComponents := 4
 		data := StbImage load(filename, x&, y&, imageComponents&, requiredComponents)
+		if (data == null) {
```
Comment (emilwestergren) [inline source/draw/RasterRgba.ooc]: Same here

**Comment by emilwestergren** [comment]
Looks good.

File: test/draw/RasterRgbaTest.ooc [hunk 4]
```diff
@@ -16,6 +16,15 @@ RasterRgbaTest: class extends Fixture {
 	sourceFlower := "test/draw/input/Flower.png"
 	init: func {
 		super("RasterRgbaTest")
+		this add("Test file not found", func {
+			try {
+				image := RasterRgba open("nonExistingFile")
+				expect(false)
```
Comment (marcusnaslund) [inline test/draw/RasterRgbaTest.ooc]: Why is this line here?

File: source/draw/RasterRgba.ooc [hunk 5]
```diff
@@ -153,6 +153,8 @@ RasterRgba: class extends RasterPacked {
 		x, y, imageComponents: Int
 		requiredComponents := 4
 		data := StbImage load(filename, x&, y&, imageComponents&, requiredComponents)
+		if (data == null)
+			Exception new("Failed to load image: " + filename) throw()
```
Comment (marcusnaslund) [inline source/draw/RasterRgba.ooc]: `raise("Failed to load image: " + filename)` is cleaner

Comment (sebastianbaginski) [inline source/draw/RasterRgba.ooc]: We could also throw directly from `StbImage load` and avoid some duplicated code.

File: test/draw/RasterRgbaTest.ooc [hunk 6]
```diff
@@ -16,6 +16,15 @@ RasterRgbaTest: class extends Fixture {
...
+				expect(image, is equal to (null as RasterRgba))
+			} catch (e: Exception) {
+				expect(e message contains("Failed to load image: "))
```
Comment (sebastianbaginski) [inline test/draw/RasterRgbaTest.ooc]: I'd rather avoid testing for the specific error message, I think it would be enough just to expect an exception.

File: source/draw/RasterRgba.ooc (hunk 5 above)
Comment (marcusnaslund) [inline source/draw/RasterRgba.ooc]: Very good, let's do that.

**Comment by marcusnaslund** [comment]
Put on hold until comments are resolved. You may add the `final review` flag again once that's done.

File: test/draw/RasterRgbaTest.ooc (hunk 6 above)
Comment (erikhagglund) [inline test/draw/RasterRgbaTest.ooc]: I agree that testing for a specific string isn't optimal, but to test that an exception is triggered will trigger on any type of exception and not necessary due to the failure we want to test. Ideally I guess you would want a separate exception type and trigger on that

File: test/draw/RasterRgbaTest.ooc (hunk 4 above)
Comment (erikhagglund) [inline test/draw/RasterRgbaTest.ooc]: I guess a comment would be good to add here, that line is there since that code should be unreachable since we are expecting an exception to trigger on the line above. So fi that row is executed the test should fail. (The row below is due to that the compiler says the variable Image is "unused" and I'm not sure to what level the compiler optimizes. I don't want the critical part of the test to be optimized away)

File: test/draw/RasterRgbaTest.ooc (hunk 6 above)
Comment (sebastianbaginski) [inline test/draw/RasterRgbaTest.ooc]: OK, we already have some Exception subtypes, maybe we can update it later.

File: test/draw/RasterRgbaTest.ooc (hunk 4 above)
Comment (sebastianbaginski) [inline test/draw/RasterRgbaTest.ooc]: >  I don't want the critical part of the test to be optimized away

I don't think gcc will completely remove a call to external function with possible side effects even with -O3, but I think you are right, we should remove this warning.

Comment (marcusnaslund) [inline test/draw/RasterRgbaTest.ooc]: Couldn't you just declare the `image` before the `try-block` and define it as `null`?

Comment (marcusnaslund) [inline test/draw/RasterRgbaTest.ooc]: Now I understand, thanks. The above comment applies to the line below.

File: test/draw/RasterRgbaTest.ooc [hunk 7]
```diff
@@ -16,6 +16,15 @@ RasterRgbaTest: class extends Fixture {
 	sourceFlower := "test/draw/input/Flower.png"
 	init: func {
 		super("RasterRgbaTest")
+		this add("Test file not found", func {
+			try {
+				image := RasterRgba open("nonExistingFile")
+				expect(false) // Unreachable code since we are expecting an exception
+				expect(image, is equal to (null as RasterRgba))
```
Comment (marcusnaslund) [inline test/draw/RasterRgbaTest.ooc]: `expect(image, is Null)` works now.
